### ISSUE TITLE
Clone method integrity checking

### ISSUE THREAD

**Issue by LeeTurl** [fo-dicom/fo-dicom#754]
### Expected behavior
When using the clone method - the length of the pixel data needs to be compared with the Rows x Columns x BitsAllocated / 8) and an exception should be thrown if the values are not equal.

Enhance Dicom needs to take into account eh number of frames also, whilst compressed data needs to be decompressed and checked.

### Actual behavior

The buffer of the new pixel data when using the clone method is calculate using the tags. This could clause data to be lost or corrupted when using clone to generate a new dataset.

Whilst I can get around the issue by checking the uncompressed pixel datasize by getting the pixel data tag and checking the length, this is not possible with compressed data.

### Steps to reproduce the behavior
Modify the tags then use the clone method

### fo-dicom version and OS/platform
4.0.0.0 on Windows 10

**Comment by mia-san** [comment]
I've just checked the DicomDataset.Clone behaviour as following.
dicom.dcm is a JPEG-compressed DICOM file with 512 x 512, 16 bits per pixel.
```
var file = DicomFile.Open("dicom.dcm");

var cols = file.Dataset.GetSingleValue<ushort>(DicomTag.Columns);
file.Dataset.AddOrUpdate(DicomTag.Columns, (ushort)(cols + 1));
Console.WriteLine("Original Columns: {0} -> {1}", cols, cols + 1);

var rows = file.Dataset.GetSingleValue<ushort>(DicomTag.Rows);
file.Dataset.AddOrUpdate(DicomTag.Rows, (ushort)(rows + 1));
Console.WriteLine("Original Rows: {0} -> {1}", rows, rows + 1);

var pd1 = DicomPixelData.Create(file.Dataset);
Console.WriteLine("Original Pixel Data Length: {0}", pd1.GetFrame(0).Size);

var cloned = file.Dataset.Clone();
Console.WriteLine("Cloned Columns: {0}", cloned.GetSingleValue<ushort>(DicomTag.Columns));
Console.WriteLine("Cloned Rows: {0}", cloned.GetSingleValue<ushort>(DicomTag.Rows));

var pd2 = DicomPixelData.Create(cloned);
Console.WriteLine("Cloned Pixel Data Length: {0}", pd2.GetFrame(0).Size);
```
outputs
```
Original Columns: 512 -> 513
Original Rows: 512 -> 513
Original Pixel Data Length: 7536
Cloned Columns: 513
Cloned Rows: 513
Cloned Pixel Data Length: 7536
```

As you see, Columns and Rows are changed, though frame size is same before and after Clone.

Or you mean something else?

**Comment by LeeTurl** [comment]
I've attempted to modify your example to show better the issue.

```
var original = DicomFile.Open("ExplicitVRLittleEndian.dcm");

var pd1 = DicomPixelData.Create(original.Dataset);

Console.WriteLine("Original Columns: {0}", pd1.Width);
Console.WriteLine("Original Rows: {0}", pd1.Height);

Console.WriteLine("Expected Pixel Data Length: {0}", pd1.BytesAllocated * pd1.Height * pd1.Width);

var pixelDataSize = pd1.GetFrame(0).Size;
Console.WriteLine("Actual Pixel Data Length: {0}", pixelDataSize);

pd1.Height -= 200;
pd1.Width -= 200;

Console.WriteLine("Modified Columns: {0}", pd1.Width);
Console.WriteLine("Modified Rows: {0}", pd1.Height);

pixelDataSize = pd1.GetFrame(0).Size;
Console.WriteLine("Actual Pixel Data Length after Rows and Columns Changed: {0}", pixelDataSize);
      
Console.WriteLine("Changing the value of Rows and Columns should not change the actual pixel data size ");
        
Console.ReadLine();
```
This causes issues because it is important that the Pixel Data size matches the header values.

Using Clone with changing the transfer syntax (Say from a compressed transfer syntax to an uncompressed on) creates a buffer based on the header values rather then decompressing the pixel data and then checking if the values match.

**Comment by mia-san** [comment]
sorry, still unclear to me what you're expecting.
you mean Rows and Columns in DicomDataset should be changed accordingly when changing DicomPixelData's property ?

**Comment by LeeTurl** [comment]
No worries, it is difficult to convey.

When cloning a dataset the original PixelData should be preserved. 

If the Tags for Rows and Columns have been changed then the cloned Pixel Data will be sized according to those tags, not the original Pixel Data.

This is important when Cloning that involves decompression, because the pixel buffer is sized incorrectly.

**Comment by mia-san** [comment]
I'm still not sure, but what about following ?
```
var original = DicomFile.Open("dicom.dcm");

var pd1 = DicomPixelData.Create(original.Dataset);
Console.WriteLine("Original Columns: {0}", pd1.Width);
Console.WriteLine("Original Rows: {0}", pd1.Height);
Console.WriteLine("Original Uncompressed Data Length: {0}", pd1.UncompressedFrameSize);
Console.WriteLine("Original Actual Data Length: {0}", pd1.GetFrame(0).Size);

var cloned = original.Dataset.Clone();
var pd2 = DicomPixelData.Create(cloned, true);

pd2.Height -= 200;
pd2.Width -= 200;
Console.WriteLine("Modified Columns: {0}", pd2.Width);
Console.WriteLine("Modified Rows: {0}", pd2.Height);
// at this point, pd2 is correlated to cloned dataset and has no frame(pixel data).
// thus pd2.GetFrame(0) throws IndexOutOfRangeException.
// use AddFrame to add frame data.
Console.WriteLine("Modified Uncompressed Data Length: {0}", pd2.UncompressedFrameSize);

//  original pixel data is unchanged.
Console.WriteLine("Original Uncompressed Data Length: {0}", pd1.UncompressedFrameSize);
Console.WriteLine("Original Actual Data Length: {0}", pd1.GetFrame(0).Size);
```

DicomPixelData.Create has second bool parameter, which creates new empty pixel data.

**Comment by LeeTurl** [comment]
**Goal**

Use Clone(TransferSyntax.ExplicitVRLittleEndian) to convert all source images regardless of transfer syntax to an uncompressed format before writing the pixel data to another stream.

Clone uses the Column, Row and BitsAllocated from the source DICOM file to calculate the uncompressed buffer size of the target PixelData and then reads from the source data copying only upto the size of the calculated buffer.

If these values are incorrect, or have been modified then the resulting target PixelData buffer will be larger or smaller than required resulting in dataloss.

The source Pixel Data should be copied into a dynamically sized buffer rather than a fixed size buffer based on the tags in the header. 

In this situation the tags in the header would be inconsistent with the pixel data, but this could be easily detected by comparing the target pixel data buffer size with (Columns x Rows * BitsAllocated / 8)

**Comment by mia-san** [comment]
I don't see any inconsistency on your program's output.
400x400x1x2 = 320,000, 800x800x1x2 = 1,280,000, which are correct.

perhaps you mean Columns, Rows etc should be checked against JPEG's SOFn marker or such on Clone?
if Columns, Rows etc don't agree with with actual pixel data, the dataset is considered to be broken.

Are you're trying to resize image or fix incorrect sized image?

**Comment by LeeTurl** [comment]
Yes, the main driver is to be able to identify when a file has a mismatch between the pixel data and the tags in the header. Under the current implementation the tags are used determine the size of the pixel data on clone. 

My proposal suggests not to do this and use a dynamic buffer for the pixel data when using clone.

Then either a check can be implemented in fellow oak or the client can determine this by checking the tags and the pixel data to see if they are consistent. 

Its all about making sure data is not lost when changing transfer syntax using the Clone method.

**Comment by mia-san** [comment]
if DICOM dataset and pixel data disagree, the dataset is considered 'broken', as I said.
I understand that repeating same data in multiple places is not good from 'single source of truth' point of view, through DICOM is designed that way as it is now.

so Clone()-ing such dataset is just a gabarge-in, garbage-out.

if you're trying to fix or detect inconsistency of such broken dataset, I suggest check those attributes as you want BEFORE Clone()-ing.
perhaps almost 'compressed' pixel data would be in JPEG(baseline|lossless|LS|2000), it should not be too difficult to know what the image size is, as format is well-known.

if you're trying to resize DICOM image, create new pixel data, fill it with resized pixel data and change tags accordingly.

**Comment by gofal** [comment]
How to classify this issue? Is it actually a bug, would it be an enhancement that fo-dicom could provide, or was it a qustion?

**Comment by mia-san** [comment]
it is not a bug, IMHO.

OP seems to request fo-dicom to check Columns, Rows etc are consistent against actual pixel data's width, height etc on Clone().
I'd say if dataset and pixel data doesn't agree, the dataset is 'corrupted'.

is it practical to provide such integrity check ?
